cpp/src/arrow/filesystem/filesystem_utils.cc

```diff
@@ -0,0 +1,203 @@
+#include "arrow/filesystem/path_util.h"
+#include "arrow/util/uri.h"
+
+namespace arrow {
+
+namespace fs {
+
+namespace factory {
+
+namespace internal {
+
+class PathInfo {
+ public:
+  PathInfo() {}
+  ~PathInfo() {}
+
+  static Status Make(const std::string& full_path, std::shared_ptr<PathInfo>* path_info) {
+    *path_info = std::make_shared<PathInfo>();
+    RETURN_NOT_OK((*path_info)->Init(full_path));
+    return Status::OK();
+  }
+
+  Status Init(const std::string& full_path) {
+    RETURN_NOT_OK(ParseURI(full_path));
+    full_path_ = std::string(full_path);
+    return Status::OK();
+  }
+
+  FileSystemType GetFileSystemType() { return fs_type_; }
+
+  std::string GetHostName() {
+    auto search = options_.find("host_name");
+    if (search == options_.end()) {
+      return "";
+    }
+    return search->second;
+  }
+
+  int GetHostPort() {
+    auto search = options_.find("host_port");
+    if (search == options_.end()) {
+      return -1;
+    }
+    std::string port_text = search->second;
+    return std::stoi(port_text);
+  }
+
+  std::string GetUser() {
+    auto search = options_.find("user");
+    if (search != options_.end()) {
+      return search->second;
+    }
+    return "";
+  }
+
+  bool GetIfUseHdfs3() {
+    auto search = options_.find("use_hdfs3");
+    if (search != options_.end()) {
+      if (search->second.compare("1") == 0) {
+        return true;
+      } else {
+        return false;
+      }
+    } else {
+      return false;
+    }
+  }
+
+  int GetRepsNum() {
+    auto search = options_.find("replication");
+    if (search != options_.end()) {
+      if (search->second.empty()) {
+        return 3;
+      } else {
+        return std::stoi(search->second);
+      }
+    } else {
+      return 3;
+    }
+  }
+
+ private:
+  FileSystemType GetFileSystemTypeFromString(const std::string& s) {
+    if (s == "hdfs") {
+      return FileSystemType::HDFS;
+    }
+    if (s == "http") {
+      return FileSystemType::S3;
+    }
+    if (s == "https") {
+      return FileSystemType::S3;
+    }
+    if (s == "file") {
+      return FileSystemType::LOCAL;
+    }
+    if (s.empty()) {
+      return FileSystemType::LOCAL;
+    }
+    return FileSystemType::UNKNOWN;
+  }
+
+  Status ParseURI(const std::string& s) {
+    arrow::internal::Uri uri;
+    RETURN_NOT_OK(uri.Parse(s));
+    fs_type_ = GetFileSystemTypeFromString(uri.scheme());
+    switch (fs_type_) {
+      case FileSystemType::HDFS:
+        options_.emplace("host_name", uri.host());
+        options_.emplace("host_port", uri.port_text());
+        RETURN_NOT_OK(ParseOptions(&uri));
+        break;
+      case FileSystemType::LOCAL:
+        RETURN_NOT_OK(ParseOptions(&uri));
+        break;
+      case FileSystemType::S3:
+        return Status::NotImplemented("S3 is not supported yet.");
+      default:
+        break;
+    }
+    return Status::OK();
+  }
+
+  Status ParseOptions(arrow::internal::Uri* uri) {
+    auto options = uri->query_items();
```

Comment:
Need to return on error, using `ARROW_ASSIGN_OR_RAISE(auto options, uri->query_items());`